helios-testing/src/main/java/com/spotify/helios/testing/HeliosSoloLogService.java

```diff
@@ -0,0 +1,111 @@
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.spotify.docker.client.DockerClient.LogsParam.follow;
+import static com.spotify.docker.client.DockerClient.LogsParam.stderr;
+import static com.spotify.docker.client.DockerClient.LogsParam.stdout;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import com.spotify.docker.client.DockerClient;
+import com.spotify.docker.client.LogStream;
+import com.spotify.helios.common.descriptors.HostStatus;
+import com.spotify.helios.common.descriptors.JobId;
+import com.spotify.helios.common.descriptors.TaskStatus;
+
+import com.google.common.collect.Maps;
+import com.google.common.util.concurrent.AbstractScheduledService;
+import com.oracle.tools.packager.Log;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.util.Map;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+class HeliosSoloLogService extends AbstractScheduledService {
+
+  private static final Logger log = LoggerFactory.getLogger(HeliosSoloLogService.class);
+
+  private final HeliosSoloDeployment solo;
+  private final DockerClient dockerClient;
+
+  private Map<String, ScheduledFuture<?>> logFutures = Maps.newHashMap();
+
+  HeliosSoloLogService(final HeliosSoloDeployment solo, final DockerClient dockerClient) {
+    this.solo = solo;
+    this.dockerClient = dockerClient;
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception {
+    try {
+      // fetch all the jobs running on the solo deployment
+      final HostStatus hostStatus = solo.client().hostStatus(solo.agentName()).get();
+      final Map<JobId, TaskStatus> statuses = hostStatus.getStatuses();
+
+      for (final TaskStatus status : statuses.values()) {
+        final String containerId = status.getContainerId();
+        if (isNullOrEmpty(containerId) || logFutures.containsKey(containerId)) {
+          continue;
+        }
+
+        // for any containers we're not already tracking, schedule attaching to their stdout/stderr
+        final ScheduledFuture<?> future = this.executor().scheduleAtFixedRate(new Runnable() {
+          @Override
+          public void run() {
+            try {
+              final LogStream logStream = dockerClient.logs(containerId, stdout(), stderr(),
+                                                            follow());
+              log.info("attaching stdout/stderr for job={}, container={}", status.getJob().getId(),
+                       containerId);
+              logStream.attach(System.out, System.err);
+            } catch (Exception e) {
+              // do nothing, we'll retry in a bit
+              log.debug("error streaming logs for job={}, container={} - {}",
+                        status.getJob().getId(), containerId, e);
+            }
+          }
+        }, 0, 100, TimeUnit.MILLISECONDS);
```

Comment:
why does attaching to the container need to be scheduled at a fix rate? won't this cause the attaching to happen many times for each container (and log the INFO message above too)?